src/transport/TCP.cpp

```diff
@@ -0,0 +1,585 @@
+    // does NOT need to be re-established.
+    for (size_t i = 0; i < mPendingPacketsSize; i++)
+    {
+        if (mPendingPackets[i].packetBuffer == nullptr)
+        {
+            if (packet == nullptr)
+            {
+                // found a slot to store the packet into
+                packet = mPendingPackets + i;
+            }
+        }
+        else if (mPendingPackets[i].peerAddress == addr)
+        {
+            // same destination exists.
+            alreadyConnecting = true;
+
+            // ensure packets are ORDERED
+            if (packet != nullptr)
+            {
+                packet->peerAddress             = addr;
+                packet->packetBuffer            = mPendingPackets[i].packetBuffer;
+                mPendingPackets[i].packetBuffer = nullptr;
+                packet                          = mPendingPackets + i;
+            }
+        }
+    }
+
+    VerifyOrExit(packet != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    // If already connecting, buffer was just enqueued for more sending
+    VerifyOrExit(!alreadyConnecting, err = CHIP_NO_ERROR);
+
+    // Ensures sufficient active connections size exist
+    VerifyOrExit(mUsedEndPointCount < mActiveConnectionsSize, err = CHIP_ERROR_NO_MEMORY);
+
+    err = mListenSocket->Layer().NewTCPEndPoint(&endPoint);
+    SuccessOrExit(err);
+
+    endPoint->AppState             = reinterpret_cast<void *>(this);
+    endPoint->OnDataReceived       = OnTcpReceive;
+    endPoint->OnConnectComplete    = OnConnectionComplete;
+    endPoint->OnConnectionClosed   = OnConnectionClosed;
+    endPoint->OnConnectionReceived = OnConnectionRecevied;
+    endPoint->OnAcceptError        = OnAcceptError;
+    endPoint->OnPeerClose          = OnPeerClosed;
+
+    err = endPoint->Connect(addr.GetIPAddress(), addr.GetPort(), addr.GetInterface());
+    SuccessOrExit(err);
+
+    // enqueue the packet once the connection succeeds
+    packet->peerAddress  = addr;
+    packet->packetBuffer = msg;
+    msg                  = nullptr;
+    mUsedEndPointCount++;
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        if (msg != nullptr)
+        {
+            System::PacketBuffer::Free(msg);
+            msg = nullptr;
+        }
+        if (endPoint != nullptr)
+        {
+            endPoint->Free();
+        }
+    }
+    return err;
+}
+
+CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, System::PacketBuffer * buffer,
+                                                       uint16_t messageSize)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    uint8_t * oldStart = buffer->Start();
+    size_t oldLength   = buffer->DataLength();
+
+    buffer->SetDataLength(messageSize);
+
+    size_t headerSize = 0;
+
+    MessageHeader header;
+    err = header.Decode(buffer->Start(), buffer->DataLength(), &headerSize);
+    SuccessOrExit(err);
+
+    buffer->ConsumeHead(headerSize);
+    HandleMessageReceived(header, peerAddress, buffer);
```

Comment:
:( This needs review then, thank you for spotting.

I don't think we can free the message as is since TCP may contain multiple messages. I will update comments and add an extra retain.

Reply:
I changed the code to AddRef, assuming the handle will do a free.